### PR TITLE
add Snappy message compression support

### PR DESCRIPTION
**Pull request by mattnworb**
Hi, this PR aims to add support for message compression with [Snappy](https://github.com/google/snappy). It is inspired by my question/request in https://github.com/grpc/grpc/issues/25390, and Snappy support seems like it was desired at points in the past (see https://github.com/grpc/grpc/issues/3384 for instance).  

---

We use tensorflow-serving for serving Machine Learning model predictions, which in our environment involves sending 100s of KBs of input data to tensorflow-serving's gRPC C++ server to make a prediction. In some cases, we currently compress messages with gzip to reduce the amount of network traffic, but we would be love to be able to use other algorithms like Snappy which can compress and decompress at faster speeds with the tradeoff of a smaller compressed size. For many of the models we use with tensorflow-serving, reducing the latency of the prediction requests by several or tens of milliseconds can be impactful for users.

Since Snappy support in grpc core (and later tensorflow-serving) is something we would benefit from, I thought it would be fun to attempt to implement it myself 😄  

--- 

In its current state, this change is incomplete. This is the first time I've attempted to contribute to a C++ project, and I have never used CMake before, let alone adding third-party dependencies to an existing Bazel or CMake build system like in what is used here. I think I'm at the point where I have some of the basics down but could use additional help (and where I should check to see if there is any desire to accept a change like this). 

If the change here sounds good in spirit, I could use pointers in a few areas to get this to a completed state:

- integrating Snappy into grpc's CMake build is incomplete here - since Snappy uses CMake too, is there a way to reuse its build completely?

- I didn't tackle stream compression with Snappy since stream compression is marked as experimental 

- in `src/core/lib/compression/message_compress.cc`, perhaps there is a simpler way to pass the data to be compressed/decompressed into `snappy::Compress` and `snappy::Decompress`

- I assume there would be additional work to do to make Snappy visible to the language bindings, such as adding `grpc.Compression.Snappy` to Python, etc

With what I have here as a start, I'm able to run `bazel run --config=dbg //examples/cpp/compression:compression_server` (with patches to use `GRPC_COMPRESS_SNAPPY` instead of `GRPC_COMPRESS_GZIP`) and can see that the server's response headers advertise support for `snappy` via `grpcurl -v`.

@drfloob

### CODE REVIEW

**Comment by veblush** [comment]
@mattnworb Thank you for the PR and we might need to discuss whether this can be added. Also adding Snappy to gRPC's dependency usually need more careful consideration because it increases the footprint of the library and Snappy should need to support all platforms gRPC supports.

**Comment by veblush** [comment]
To enable you to see the CI test result while working on the PR, I just added the label to kick off CI tests.

File: include/grpc/impl/codegen/compression_types.h [hunk 1]
```diff
@@ -58,9 +58,9 @@ typedef enum {
   GRPC_COMPRESS_NONE = 0,
   GRPC_COMPRESS_DEFLATE,
   GRPC_COMPRESS_GZIP,
+  GRPC_COMPRESS_SNAPPY,
```
Comment (mattnworb) [inline include/grpc/impl/codegen/compression_types.h:61]: a question I had here: is the int value of an enum option like `GRPC_COMPRESS_STREAM_GZIP` considered a part of the public API?  I wasn't sure if it is a bad practice to insert a new enum value in the middle of possible values here like this, which causes the int value of `GRPC_COMPRESS_STREAM_GZIP` to move up by one. 

For instance https://github.com/grpc/grpc/blob/master/doc/compression_cookbook.md#disable-compression-algorithms mentions using these int values to compute a bitset to set channel arguments like GRPC_COMPRESSION_CHANNEL_ENABLED_ALGORITHMS_BITSET

Adding `GRPC_COMPRESS_SNAPPY` here allowed the implementation of `grpc_compression_bitset_to_message_bitset` etc to remain unchanged,

**Comment by jtattermusch** [comment]
I think we need to think twice before we accept this. You need to have a really good reason to increase the footprint of grpc and to add a new dependency (and there's a lot of maintenance associated with adding new dependencies).
Is there a solution that would allow you to register snappy compression with grpc dynamically?

**Comment by mattnworb** [comment]
@jtattermusch I think that makes sense, I can certainly understand there are a lot of after-effects when adding new dependencies here (`third_party/README.md` makes this pretty clear).

But is there anything already existing within the compression support here that would allow for loading support for certain algorithms dynamically? I didn't see anything like that, but I am also very new to this codebase (and C/C++ in general).

**Comment by veblush** [comment]
@mattnworb We had a discussion internally and drew a conclusion that we cannot get this change because this should be added to gRPC Java and Go but it's not feasible. Instead, we'd better to pursue the path to support it using user-defined compression algorithm which doesn't require gRPC to have it by default. gRPC Java and Go support it so it'd make more sense to have the same approach for gRPC Core & variants.

**Comment by mattnworb** [comment]
@veblush thanks for the update. 

> Instead, we'd better to pursue the path to support it using user-defined compression algorithm which doesn't require gRPC to have it by default

Would this mean something like: I can start up a grpc C++ server with some (optional) configuration to enable message compression/decompression with some arbitrary library (e.g. `ServerBuilder.enableMessageCompression("libsnappy")`), as long as the library implements a certain interface? I think https://github.com/grpc/grpc/issues/25156 kind of suggests this.

If so, is this something you'd be open to collaborating on? I'd be interesting in helping make this functionality a reality. Are there any examples today of anything in the core grpc project where support for an optional feature loads a library that implements the functionality dynamically?

**Comment by veblush** [comment]
@mattnworb Yes, you're right. We haven't worked on this yet and it might need some design work on both C Core and C++. Since C Core does actual compression job, C Core also needs changes while supporting other wrapped languages.

**Pull request closed without merging**